<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>Accessing EuPathDB Resources using AnnotationHub</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Accessing EuPathDB Resources using AnnotationHub</h1>



<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>
<p><strong>Authors</strong>: <a href="mailto:keith.hughitt@nih.gov">V. Keith Hughitt</a><br /> <strong>Modified:</strong> NA<br /> <strong>Compiled</strong>: Tue May 21 11:49:18 2019</p>
<div id="overview" class="section level1">
<h1>Overview</h1>
<p>This tutorial describes how to query and make use of annotations retrieved from <a href="http://eupathdb.org/eupathdb/">EuPathDB : The Eukaryotic Pathogen Genomics Resource</a> using <a href="http://bioconductor.org/packages/release/bioc/html/AnnotationHub.html">AnnotationHub</a>.</p>
<p>For more information on using AnnotationHub, check out the AnnotationHub vignettes:</p>
<ul>
<li><a href="http://bioconductor.org/packages/release/bioc/vignettes/AnnotationHub/inst/doc/AnnotationHub-HOWTO.html">AnnotationHub: Access the AnnotationHub Web Service</a></li>
<li><a href="http://bioconductor.org/packages/release/bioc/vignettes/AnnotationHub/inst/doc/AnnotationHub-HOWTO.html">AnnotationHub How-To’s</a></li>
</ul>
<p>The resources described in this tutorial were generating using GFF files and web API requests made to the various EuPathDB databases (TriTrypDB, ToxoDB, etc.) Only organisms with annotated genomes (those for which GFF files are available) are accessible through AnnotationHub.</p>
<p>The two main resources provided are:</p>
<ul>
<li><a href="https://www.bioconductor.org/help/workflows/annotation/annotation/#OrgDb">OrgDb</a></li>
<li><a href="http://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GRanges</a></li>
</ul>
<p>OrgDb objects for an organism include basic gene-level information such as:</p>
<ul>
<li>Gene ID</li>
<li>Gene description</li>
<li>Chromosome number</li>
<li>GO terms associated with gene</li>
<li>KEGG Pathways associated with gene</li>
<li>Etc.</li>
</ul>
<p>For some organisms, <a href="https://www.ebi.ac.uk/interpro/">InterPro</a> protein domain information is also available (in some cases, however, even though InterPro domain information is available through EuPathDB, it is too large to be included in the current AnnotationHub resources).</p>
<p>For more information about working with Bioconductor annotation resources, see:</p>
<ul>
<li><a href="https://www.bioconductor.org/help/workflows/annotation/annotation/">Genomic Annotation Resources in Bioconductor</a></li>
</ul>
</div>
<div id="installation" class="section level1">
<h1>Installation</h1>
<p>If you don’t already have AnnotationHub installed on your system, use <code>BiocManager::install</code> to install the package:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</a>
<a class="sourceLine" id="cb1-2" title="2">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;AnnotationHub&quot;</span>)</a></code></pre></div>
</div>
<div id="getting-started" class="section level1">
<h1>Getting started</h1>
<p>To begin, let’s create a new <code>AnnotationHub</code> connection and use it to query AnnotationHub for all EuPathDB resources.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="kw">library</span>(<span class="st">&quot;EuPathDB&quot;</span>)</a></code></pre></div>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: 'BiocGenerics'</code></pre>
<pre><code>## The following objects are masked from 'package:parallel':
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     anyDuplicated, append, as.data.frame, basename, cbind,
##     colnames, dirname, do.call, duplicated, eval, evalq, Filter,
##     Find, get, grep, grepl, intersect, is.unsorted, lapply, Map,
##     mapply, match, mget, order, paste, pmax, pmax.int, pmin,
##     pmin.int, Position, rank, rbind, Reduce, rownames, sapply,
##     setdiff, sort, table, tapply, union, unique, unsplit, which,
##     which.max, which.min</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     'browseVignettes()'. To cite Bioconductor, see
##     'citation(&quot;Biobase&quot;)', and for packages 'citation(&quot;pkgname&quot;)'.</code></pre>
<pre><code>## Loading required package: GenomicRanges</code></pre>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## Loading required package: S4Vectors</code></pre>
<pre><code>## 
## Attaching package: 'S4Vectors'</code></pre>
<pre><code>## The following object is masked from 'package:base':
## 
##     expand.grid</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## Loading required package: GenomeInfoDb</code></pre>
<pre><code>## Loading required package: GenomeInfoDbData</code></pre>
<pre><code>## Loading required package: AnnotationHub</code></pre>
<pre><code>## Loading required package: BiocFileCache</code></pre>
<pre><code>## Loading required package: dbplyr</code></pre>
<pre><code>## 
## Attaching package: 'AnnotationHub'</code></pre>
<pre><code>## The following object is masked from 'package:Biobase':
## 
##     cache</code></pre>
<pre><code>## 
## This is EuPathDB version 1.5.0
##  Read 'EuPathDB()' to get started.</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" title="1"><span class="kw">library</span>(<span class="st">&quot;AnnotationHub&quot;</span>)</a>
<a class="sourceLine" id="cb25-2" title="2"></a>
<a class="sourceLine" id="cb25-3" title="3"><span class="co"># create an AnnotationHub connection</span></a>
<a class="sourceLine" id="cb25-4" title="4">ah &lt;-<span class="st"> </span><span class="kw">AnnotationHub</span>()</a></code></pre></div>
<pre><code>## snapshotDate(): 2019-05-02</code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" title="1"><span class="co"># search for all EuPathDB resources</span></a>
<a class="sourceLine" id="cb27-2" title="2">meta &lt;-<span class="st"> </span><span class="kw">query</span>(ah, <span class="st">&quot;EuPathDB&quot;</span>)</a>
<a class="sourceLine" id="cb27-3" title="3"></a>
<a class="sourceLine" id="cb27-4" title="4"><span class="kw">length</span>(meta)</a></code></pre></div>
<pre><code>## [1] 1790</code></pre>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" title="1"><span class="kw">head</span>(meta)</a></code></pre></div>
<pre><code>## AnnotationHub with 6 records
## # snapshotDate(): 2019-05-02 
## # $dataprovider: FungiDB, TriTrypDB, MicrosporidiaDB
## # $species: Aspergillus sydowii, Aspergillus versicolor, Edhazardia aed...
## # $rdataclass: GRanges
## # additional mcols(): taxonomyid, genome, description,
## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass,
## #   tags, rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., 'object[[&quot;AH65266&quot;]]' 
## 
##             title                                                         
##   AH65266 | Edhazardia aedis USNM 41457 transcript information            
##   AH65267 | Aspergillus versicolor CBS 583.65 transcript information      
##   AH65268 | Aspergillus sydowii CBS 593.65 transcript information         
##   AH65269 | Phytophthora cinnamomi var. cinnamomi CBS 144.22 transcript...
##   AH65270 | Trypanosoma vivax Y486 transcript information                 
##   AH65271 | Trypanosoma congolense IL3000 transcript information</code></pre>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" title="1"><span class="co"># types of EuPathDB data available</span></a>
<a class="sourceLine" id="cb31-2" title="2"><span class="kw">table</span>(meta<span class="op">$</span>rdataclass)</a></code></pre></div>
<pre><code>## 
## GRanges   OrgDb 
##    1028     762</code></pre>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" title="1"><span class="co"># distribution of resources by specific databases</span></a>
<a class="sourceLine" id="cb33-2" title="2"><span class="kw">table</span>(meta<span class="op">$</span>dataprovider)</a></code></pre></div>
<pre><code>## 
##        AmoebaDB        CryptoDB         FungiDB       GiardiaDB 
##             191              81             669              25 
## MicrosporidiaDB    PiroplasmaDB        PlasmoDB       SchistoDB 
##             170              49             195               6 
##          ToxoDB         TrichDB       TriTrypDB 
##             154               5             245</code></pre>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" title="1"><span class="co"># list of organisms for which resources are available</span></a>
<a class="sourceLine" id="cb35-2" title="2"><span class="kw">length</span>(<span class="kw">unique</span>(meta<span class="op">$</span>species))</a></code></pre></div>
<pre><code>## [1] 448</code></pre>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb37-1" title="1"><span class="kw">head</span>(<span class="kw">unique</span>(meta<span class="op">$</span>species))</a></code></pre></div>
<pre><code>## [1] &quot;Edhazardia aedis&quot;       &quot;Aspergillus versicolor&quot;
## [3] &quot;Aspergillus sydowii&quot;    &quot;Phytophthora cinnamomi&quot;
## [5] &quot;Trypanosoma vivax&quot;      &quot;Trypanosoma congolense&quot;</code></pre>
</div>
<div id="working-with-eupathdb-orgdb-resources" class="section level1">
<h1>Working with EuPathDB OrgDb resources</h1>
<p>Next, we will see how you can query AnnotationHub for EuPathDB OrgDb resources.</p>
<p>To begin, create an AnnotationHub connection, if you have not already done so, as shown in the section above.</p>
<p>You can now use the <code>query</code> function to search for your organism of interest and store the result as follows:</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" title="1">res &lt;-<span class="st"> </span><span class="kw">query</span>(ah, <span class="kw">c</span>(<span class="st">&quot;OrgDb&quot;</span>, <span class="st">&quot;Leishmania major strain Friedlin&quot;</span>))</a>
<a class="sourceLine" id="cb39-2" title="2">res</a></code></pre></div>
<pre><code>## AnnotationHub with 2 records
## # snapshotDate(): 2019-05-02 
## # $dataprovider: TriTrypDB
## # $species: Leishmania major strain Friedlin
## # $rdataclass: OrgDb
## # additional mcols(): taxonomyid, genome, description,
## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass,
## #   tags, rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., 'object[[&quot;AH71675&quot;]]' 
## 
##             title                                               
##   AH71675 | Transcript information for Leishmania major Friedlin
##   AH72076 | Transcript information for Leishmania major Friedlin</code></pre>
<p>The result includes a single record, “AH56967”. The record can be accessed from the result variable using list-like indexing:</p>
<p>Hmm it appears that something at annotationhub is being updated and all my queries are getting returned 403, Forbidden. I will for the moment stop the following blocks from running.</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb41-1" title="1">chosen_ah &lt;-<span class="st"> </span>res[[<span class="dv">2</span>]]</a>
<a class="sourceLine" id="cb41-2" title="2">chosen_ah</a></code></pre></div>
<p>We can see that we now have an OrgDb instance, and as such, we can use the usual methods available for working this OrgDb objects, including:</p>
<ul>
<li><code>columns()</code></li>
<li><code>keys()</code></li>
<li><code>select()</code></li>
</ul>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" title="1"><span class="co"># list available fields to retrieve</span></a>
<a class="sourceLine" id="cb42-2" title="2"><span class="kw">columns</span>(orgdb)</a>
<a class="sourceLine" id="cb42-3" title="3"></a>
<a class="sourceLine" id="cb42-4" title="4"><span class="co"># create a vector containing all gene ids for the organism</span></a>
<a class="sourceLine" id="cb42-5" title="5">gids &lt;-<span class="st"> </span><span class="kw">keys</span>(orgdb, <span class="dt">keytype=</span><span class="st">&quot;GID&quot;</span>)</a>
<a class="sourceLine" id="cb42-6" title="6"><span class="kw">head</span>(gids)</a>
<a class="sourceLine" id="cb42-7" title="7"></a>
<a class="sourceLine" id="cb42-8" title="8"><span class="co"># retrieve the chromosome, description, and biotype for each gene</span></a>
<a class="sourceLine" id="cb42-9" title="9">dat &lt;-<span class="st"> </span><span class="kw">select</span>(orgdb, <span class="dt">keys=</span>gids, <span class="dt">keytype=</span><span class="st">&quot;GID&quot;</span>, <span class="dt">columns=</span><span class="kw">c</span>(<span class="st">&quot;CHR&quot;</span>, <span class="st">&quot;TYPE&quot;</span>, <span class="st">&quot;GENEDESCRIPTION&quot;</span>))</a>
<a class="sourceLine" id="cb42-10" title="10"></a>
<a class="sourceLine" id="cb42-11" title="11"><span class="kw">head</span>(dat)</a>
<a class="sourceLine" id="cb42-12" title="12"></a>
<a class="sourceLine" id="cb42-13" title="13"><span class="kw">table</span>(dat<span class="op">$</span>TYPE)</a>
<a class="sourceLine" id="cb42-14" title="14"><span class="kw">table</span>(dat<span class="op">$</span>CHR)</a>
<a class="sourceLine" id="cb42-15" title="15"></a>
<a class="sourceLine" id="cb42-16" title="16"><span class="co"># create a gene / GO term mapping</span></a>
<a class="sourceLine" id="cb42-17" title="17">gene_go_mapping &lt;-<span class="st"> </span><span class="kw">select</span>(orgdb, <span class="dt">keys=</span>gids, <span class="dt">keytype=</span><span class="st">'GID'</span>,</a>
<a class="sourceLine" id="cb42-18" title="18">                          <span class="dt">columns=</span><span class="kw">c</span>(<span class="st">&quot;GO_ID&quot;</span>, <span class="st">&quot;GO_TERM_NAME&quot;</span>, <span class="st">&quot;ONTOLOGY&quot;</span>))</a>
<a class="sourceLine" id="cb42-19" title="19"><span class="kw">head</span>(gene_go_mapping)</a>
<a class="sourceLine" id="cb42-20" title="20"></a>
<a class="sourceLine" id="cb42-21" title="21"><span class="co"># retrieve KEGG, etc. pathway annotations</span></a>
<a class="sourceLine" id="cb42-22" title="22">gene_pathway_mapping &lt;-<span class="st"> </span><span class="kw">select</span>(orgdb, <span class="dt">keys=</span>gids, <span class="dt">keytype=</span><span class="st">&quot;GID&quot;</span>,</a>
<a class="sourceLine" id="cb42-23" title="23">                               <span class="dt">columns=</span><span class="kw">c</span>(<span class="st">&quot;PATHWAY&quot;</span>, <span class="st">&quot;PATHWAY_SOURCE&quot;</span>))</a>
<a class="sourceLine" id="cb42-24" title="24"><span class="kw">table</span>(gene_pathway_mapping<span class="op">$</span>PATHWAY_SOURCE)</a>
<a class="sourceLine" id="cb42-25" title="25"><span class="kw">head</span>(gene_pathway_mapping)</a></code></pre></div>
</div>
<div id="working-with-eupathdb-granges-resources" class="section level1">
<h1>Working with EuPathDB GRanges resources</h1>
<p>In addition to retrieving gene annotations, AnnotationHub can also be used to query GenomicRange (GRange) objects containing information about gene and transcript structure.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb43-1" title="1"><span class="co"># query AnnotationHub</span></a>
<a class="sourceLine" id="cb43-2" title="2">res &lt;-<span class="st"> </span><span class="kw">query</span>(ah, <span class="kw">c</span>(<span class="st">&quot;Leishmania major strain Friedlin&quot;</span>, <span class="st">&quot;GRanges&quot;</span>, <span class="st">&quot;EuPathDB&quot;</span>))</a>
<a class="sourceLine" id="cb43-3" title="3">res</a></code></pre></div>
<pre><code>## AnnotationHub with 3 records
## # snapshotDate(): 2019-05-02 
## # $dataprovider: TriTrypDB
## # $species: Leishmania major strain Friedlin, Leishmania major
## # $rdataclass: GRanges
## # additional mcols(): taxonomyid, genome, description,
## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass,
## #   tags, rdatapath, sourceurl, sourcetype 
## # retrieve records with, e.g., 'object[[&quot;AH65354&quot;]]' 
## 
##             title                                                  
##   AH65354 | Leishmania major strain Friedlin transcript information
##   AH70912 | Transcript information for Leishmania major Friedlin   
##   AH71314 | Transcript information for Leishmania major Friedlin</code></pre>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" title="1"><span class="co"># retrieve a GRanges instance associated with the result record</span></a>
<a class="sourceLine" id="cb45-2" title="2">gr &lt;-<span class="st"> </span>res[[<span class="dv">3</span>]]</a></code></pre></div>
<pre><code>## downloading 0 resources</code></pre>
<pre><code>## loading from cache 
##     'AH71314 : 78060'</code></pre>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb48-1" title="1"><span class="kw">summary</span>(gr)</a></code></pre></div>
<pre><code>## [1] &quot;GRanges object with 37311 ranges and 10 metadata columns&quot;</code></pre>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb50-1" title="1"><span class="kw">head</span>(gr)</a></code></pre></div>
<pre><code>## GRanges object with 6 ranges and 10 metadata columns:
##       seqnames        ranges strand |   source     type     score
##          &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt;
##   [1]  LmjF.24 117748-119043      + | EuPathDB     gene      &lt;NA&gt;
##   [2]  LmjF.24 117748-119043      + | EuPathDB     mRNA      &lt;NA&gt;
##   [3]  LmjF.24 117748-119043      + | EuPathDB     exon      &lt;NA&gt;
##   [4]  LmjF.24 117748-119043      + | EuPathDB      CDS      &lt;NA&gt;
##   [5]  LmjF.34 961963-966600      - | EuPathDB     gene      &lt;NA&gt;
##   [6]  LmjF.34 961963-966600      - | EuPathDB     mRNA      &lt;NA&gt;
##           phase                        ID
##       &lt;integer&gt;               &lt;character&gt;
##   [1]      &lt;NA&gt;              LmjF.24.0370
##   [2]      &lt;NA&gt;         LmjF.24.0370:mRNA
##   [3]      &lt;NA&gt;      exon_LmjF.24.0370-E1
##   [4]         0 LmjF.24.0370:mRNA-p1-CDS1
##   [5]      &lt;NA&gt;              LmjF.34.2180
##   [6]      &lt;NA&gt;         LmjF.34.2180:mRNA
##                                description            Parent
##                                &lt;character&gt;   &lt;CharacterList&gt;
##   [1] aspartate aminotransferase, putative              &lt;NA&gt;
##   [2] aspartate aminotransferase, putative      LmjF.24.0370
##   [3]                                 &lt;NA&gt; LmjF.24.0370:mRNA
##   [4]                                 &lt;NA&gt; LmjF.24.0370:mRNA
##   [5]      hypothetical protein, conserved              &lt;NA&gt;
##   [6]      hypothetical protein, conserved      LmjF.34.2180
##                                   Note   Ontology_term
##                        &lt;CharacterList&gt; &lt;CharacterList&gt;
##   [1]                             &lt;NA&gt;            &lt;NA&gt;
##   [2] 2.6.1.1 (Aspartate transaminase)      GO_0005737
##   [3]                             &lt;NA&gt;            &lt;NA&gt;
##   [4]                             &lt;NA&gt;            &lt;NA&gt;
##   [5]                             &lt;NA&gt;            &lt;NA&gt;
##   [6]                             &lt;NA&gt;            &lt;NA&gt;
##          protein_source_id
##                &lt;character&gt;
##   [1]                 &lt;NA&gt;
##   [2]                 &lt;NA&gt;
##   [3]                 &lt;NA&gt;
##   [4] LmjF.24.0370:mRNA-p1
##   [5]                 &lt;NA&gt;
##   [6]                 &lt;NA&gt;
##   -------
##   seqinfo: 36 sequences from TriTrypDB-43_LmajorFriedlin genome; no seqlengths</code></pre>
<p>The resulting <code>GRanges</code> object can then be interacted with using the <a href="https://bioconductor.org/packages/3.7/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.pdf">standard GRanges functions</a>, including:</p>
<ul>
<li>seqnames</li>
<li>strand</li>
<li>width</li>
</ul>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb52-1" title="1"><span class="co"># chromosome names</span></a>
<a class="sourceLine" id="cb52-2" title="2"><span class="kw">seqnames</span>(gr)</a></code></pre></div>
<pre><code>## factor-Rle of length 37311 with 9006 runs
##   Lengths:       4       4       4       4 ...       4       4       4
##   Values : LmjF.24 LmjF.34 LmjF.36 LmjF.28 ... LmjF.04 LmjF.11 LmjF.27
## Levels(36): LmjF.01 LmjF.02 LmjF.03 LmjF.04 ... LmjF.34 LmjF.35 LmjF.36</code></pre>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb54-1" title="1"><span class="co"># strand information</span></a>
<a class="sourceLine" id="cb54-2" title="2"><span class="kw">strand</span>(gr)</a></code></pre></div>
<pre><code>## factor-Rle of length 37311 with 4630 runs
##   Lengths:  4 12  4  4  4  4  3  4 31  4 ...  8  8  4  4  4  4  4 23  4  4
##   Values :  +  -  +  -  +  -  +  -  +  - ...  +  -  +  -  +  -  +  -  +  -
## Levels(3): + - *</code></pre>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb56-1" title="1"><span class="co"># feature widths</span></a>
<a class="sourceLine" id="cb56-2" title="2"><span class="kw">head</span>(<span class="kw">width</span>(gr))</a></code></pre></div>
<pre><code>## [1] 1296 1296 1296 1296 4638 4638</code></pre>
<p>Some information can be retrieve directly as object properties using the <code>$</code> operator:</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb58-1" title="1"><span class="co"># list of location types in the resource</span></a>
<a class="sourceLine" id="cb58-2" title="2"><span class="kw">table</span>(gr<span class="op">$</span>type)</a></code></pre></div>
<pre><code>## 
##            gene            mRNA            exon             CDS 
##            9378            8519            9583            8606 
##          snoRNA            rRNA three_prime_UTR  five_prime_UTR 
##             741              63               8             241 
##           ncRNA            tRNA           snRNA 
##              83              83               6</code></pre>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb60-1" title="1"><span class="kw">table</span>(gr<span class="op">@</span>strand)</a></code></pre></div>
<pre><code>## 
##     +     -     * 
## 20366 16945     0</code></pre>
<p>To subset the GRanges instance, you can use the standard <code>[</code> operator:</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb62-1" title="1"><span class="co"># get the first three ranges</span></a>
<a class="sourceLine" id="cb62-2" title="2">gr[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 10 metadata columns:
##       seqnames        ranges strand |   source     type     score
##          &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt;
##   [1]  LmjF.24 117748-119043      + | EuPathDB     gene      &lt;NA&gt;
##   [2]  LmjF.24 117748-119043      + | EuPathDB     mRNA      &lt;NA&gt;
##   [3]  LmjF.24 117748-119043      + | EuPathDB     exon      &lt;NA&gt;
##           phase                   ID                          description
##       &lt;integer&gt;          &lt;character&gt;                          &lt;character&gt;
##   [1]      &lt;NA&gt;         LmjF.24.0370 aspartate aminotransferase, putative
##   [2]      &lt;NA&gt;    LmjF.24.0370:mRNA aspartate aminotransferase, putative
##   [3]      &lt;NA&gt; exon_LmjF.24.0370-E1                                 &lt;NA&gt;
##                  Parent                             Note   Ontology_term
##         &lt;CharacterList&gt;                  &lt;CharacterList&gt; &lt;CharacterList&gt;
##   [1]              &lt;NA&gt;                             &lt;NA&gt;            &lt;NA&gt;
##   [2]      LmjF.24.0370 2.6.1.1 (Aspartate transaminase)      GO_0005737
##   [3] LmjF.24.0370:mRNA                             &lt;NA&gt;            &lt;NA&gt;
##       protein_source_id
##             &lt;character&gt;
##   [1]              &lt;NA&gt;
##   [2]              &lt;NA&gt;
##   [3]              &lt;NA&gt;
##   -------
##   seqinfo: 36 sequences from TriTrypDB-43_LmajorFriedlin genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb64-1" title="1"><span class="co"># get all gene entries on chromosome 4</span></a>
<a class="sourceLine" id="cb64-2" title="2">chr4_genes &lt;-<span class="st"> </span>gr[gr<span class="op">$</span>type <span class="op">==</span><span class="st"> 'gene'</span> <span class="op">&amp;</span><span class="st"> </span><span class="kw">seqnames</span>(gr) <span class="op">==</span><span class="st"> 'LmjF.04'</span>]</a>
<a class="sourceLine" id="cb64-3" title="3"><span class="kw">summary</span>(chr4_genes)</a></code></pre></div>
<pre><code>## [1] &quot;GRanges object with 130 ranges and 10 metadata columns&quot;</code></pre>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb66-1" title="1"><span class="co">## Hey, checkit, there are 130 genes on chromosome 4.</span></a></code></pre></div>
</div>
<div id="session-information" class="section level1">
<h1>Session Information</h1>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb67-1" title="1"><span class="kw">sessionInfo</span>()</a></code></pre></div>
<pre><code>## R version 3.6.0 (2019-04-26)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Devuan GNU/Linux beowulf/ceres
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
## LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.3.5.so
## 
## locale:
##  [1] LC_CTYPE=en_US.utf8       LC_NUMERIC=C             
##  [3] LC_TIME=en_US.utf8        LC_COLLATE=en_US.utf8    
##  [5] LC_MONETARY=en_US.utf8    LC_MESSAGES=en_US.utf8   
##  [7] LC_PAPER=en_US.utf8       LC_NAME=C                
##  [9] LC_ADDRESS=C              LC_TELEPHONE=C           
## [11] LC_MEASUREMENT=en_US.utf8 LC_IDENTIFICATION=C      
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] EuPathDB_1.5.0         AnnotationHub_2.16.0   BiocFileCache_1.8.0   
##  [4] dbplyr_1.4.0           GenomeInfoDbData_1.2.1 GenomicRanges_1.36.0  
##  [7] GenomeInfoDb_1.20.0    IRanges_2.18.0         S4Vectors_0.22.0      
## [10] Biobase_2.44.0         BiocGenerics_0.30.0    BiocStyle_2.12.0      
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.1                    prettyunits_1.0.2            
##  [3] ps_1.3.0                      foreach_1.4.4                
##  [5] assertthat_0.2.1              rprojroot_1.3-2              
##  [7] digest_0.6.19                 mime_0.6                     
##  [9] R6_2.4.0                      backports_1.1.4              
## [11] RSQLite_2.1.1                 evaluate_0.13                
## [13] httr_1.4.0                    pillar_1.4.0                 
## [15] zlibbioc_1.30.0               rlang_0.3.4                  
## [17] curl_3.3                      data.table_1.12.2            
## [19] callr_3.2.0                   blob_1.1.1                   
## [21] rmarkdown_1.12                desc_1.2.0                   
## [23] devtools_2.0.2                stringr_1.4.0                
## [25] RCurl_1.95-4.12               bit_1.1-14                   
## [27] shiny_1.3.2                   compiler_3.6.0               
## [29] httpuv_1.5.1                  xfun_0.7                     
## [31] pkgconfig_2.0.2               pkgbuild_1.0.3               
## [33] htmltools_0.3.6               tidyselect_0.2.5             
## [35] tibble_2.1.1                  interactiveDisplayBase_1.22.0
## [37] codetools_0.2-16              crayon_1.3.4                 
## [39] dplyr_0.8.1                   withr_2.1.2                  
## [41] later_0.8.0                   bitops_1.0-6                 
## [43] rappdirs_0.3.1                jsonlite_1.6                 
## [45] xtable_1.8-4                  DBI_1.0.0                    
## [47] magrittr_1.5                  cli_1.1.0                    
## [49] stringi_1.4.3                 XVector_0.24.0               
## [51] fs_1.3.1                      promises_1.0.1               
## [53] remotes_2.0.4                 testthat_2.1.1               
## [55] xml2_1.2.0                    iterators_1.0.10             
## [57] tools_3.6.0                   bit64_0.9-7                  
## [59] glue_1.3.1                    purrr_0.3.2                  
## [61] processx_3.3.1                pkgload_1.0.2                
## [63] yaml_2.2.0                    AnnotationDbi_1.46.0         
## [65] BiocManager_1.30.4            sessioninfo_1.1.1            
## [67] rvest_0.3.4                   memoise_1.1.0                
## [69] knitr_1.23                    usethis_1.5.0</code></pre>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
